File: src/main/java/com/wxl/dyttcrawler/downloader/DyttRedirectStrategy.java
package com.wxl.dyttcrawler.downloader;

import lombok.extern.slf4j.Slf4j;
import org.apache.http.*;
import org.apache.http.impl.client.LaxRedirectStrategy;
import org.apache.http.protocol.HttpContext;

import java.net.URI;

/**
 * Create by wuxingle on 2020/7/18
 * 电影天堂的重定向策略
 * 200也可能是重定向。
 */
@Slf4j
public class DyttRedirectStrategy extends LaxRedirectStrategy {


    @Override
    public boolean isRedirected(HttpRequest request, HttpResponse response, HttpContext context) throws ProtocolException {
        boolean isMoved = super.isRedirected(request, response, context);
        if (!isMoved) {
            final String method = request.getRequestLine().getMethod();
            if (isRedirectable(method)) {
                final Header locationHeader = response.getFirstHeader(HttpHeaders.CONTENT_LOCATION);
                if (locationHeader != null) {
                    log.debug("{},has content-location header:{}",
                            request.getRequestLine().getUri(), locationHeader.getValue());
                    response.addHeader("location", locationHeader.getValue());
                    return true;

                }
            }
            return false;
        }
        return true;
    }

    @Override
    public URI getLocationURI(HttpRequest request, HttpResponse response, HttpContext context) throws ProtocolException {
        return super.getLocationURI(request, response, context);
    }
}

File: src/main/kotlin/com/wxl/dyttcrawler/downloader/DyttRedirectStrategy.kt
package com.wxl.dyttcrawler.downloader

import org.apache.http.HttpHeaders
import org.apache.http.HttpRequest
import org.apache.http.HttpResponse
import org.apache.http.impl.client.LaxRedirectStrategy
import org.apache.http.protocol.HttpContext
import org.slf4j.LoggerFactory
import java.net.URI

/**
 * Create by wuxingle on 2021/10/08
 * 电影天堂的重定向策略
 * 200也可能是重定向。
 */
class DyttRedirectStrategy : LaxRedirectStrategy() {

    companion object {
        private val log = LoggerFactory.getLogger(DyttRedirectStrategy::class.java)
    }

    override fun isRedirected(request: HttpRequest, response: HttpResponse, context: HttpContext): Boolean {
        val isMoved = super.isRedirected(request, response, context)
        if (!isMoved) {
            val method = request.requestLine.method
            if (isRedirectable(method)) {
                val locationHeader = response.getFirstHeader(HttpHeaders.CONTENT_LOCATION)
                if (locationHeader != null) {
                    log.debug(
                        "{},has content-location header:{}",
                        request.requestLine.uri, locationHeader.value
                    )
                    response.addHeader("location", locationHeader.value)
                    return true
                }
            }
            return false
        }
        return true
    }

    override fun getLocationURI(request: HttpRequest, response: HttpResponse, context: HttpContext): URI {
        return super.getLocationURI(request, response, context)
    }
}